### PR TITLE
ci: force rebase and autosquash before merging

### PR DESCRIPTION
**Pull request by frazarshad**
refs: #8500

## Description

This PR does the following:
- adds a new rule to mergify which forces it to rebase and autosquash the branch. 
- adds a condition of `linear-history` to the queuing rule for rebasing, so that only rebased PRs can be queued. 
- adds a new fixup commit ci check which is a required condition for merging

**Note**: autosquash here refers to `--autosquash` flag used by the `git rebase` command. it is used to squash and combine `fixup!` commits. it does not squash all commits into a single commit before merging

It fixes the following problems in #8500 
- rebases the PR to remove all merge commits and `fixup!` commits
- forces rebase on all PRs even if they are up to date with the master branch
- enforce `linear-history` check on `automerge:rebase` enqueue to always enforce a linear-history before merging
- the `rebase and autosquash` rule in mergify also fails if the PR cannot be rebased without conflicts. The failure indicates to the user that the PR must be rebased manually

### Examples of use:
Testing rebase on PR that was up to date: https://github.com/frazarshad/mergify-experiements/pull/2
Testing if a fixup with incorrect title stops the PR from merging: https://github.com/frazarshad/agoric-sdk/pull/6

### Fallback:
In case the new action does not work as expected and does not auto-rebase, the user can still manually rebase in order to use the `automerge:rebase` action. All other `automerge` actions work as originally intended.

In order to bypass the `no-fixup-commit` check, simply add a `bypass:linear-history` label to the PR

### Comparison report:
This is a comparison made of the previous method for merging and the new method: https://drive.google.com/file/d/1kfIW-pVOORuYwBOQlxt-2cus0k9ynAfb/view?usp=drive_link

### CODE REVIEW

**Comment by mhofman** [comment]
I'm confused, does this force a rebase on every PR that is labelled `automerge:rebase`? That seems like that may be costly on PRs that don't have any fixup or merge commits.

I also don't quite understand how the added  condition for linear history helps triggering this 2, step merge process.

**Comment by warner** [comment]
My favorite history shape is a series of either non-overlapping linear branches with a single merge commit at the end:

Screenshot 2024-07-30 at 10 21 05 AM


or single commits:

Screenshot 2024-07-30 at 10 22 04 AM

A policy which always rebases, but then always uses a merge commit, makes each of those single commits into a one-commit branch (like the first case, but with one commit per PR instead of more than one), which is kind of verbose but tolerable:

Screenshot 2024-07-30 at 10 21 38 AM


We've been pretty good about maintaining these shapes.. I had to dig pretty far back to find examples of anything else. The shape that bugs me a little bit is when we get a lot of overlap between branches, like when this kind of railroad-diagram becomes more than two columns wide. That can happen when branches aren't rebased before merging:

Screenshot 2024-07-30 at 10 22 40 AM

and then the kind that I find really annoying is when we need to merge trunk into the feature branch one or more times, and/or merging secondary fixup branches (like when author A makes a PR, and author B makes a second PR to merge into A's branch before A can merge their branch into trunk). This kind of tangle is hard to pick apart later, when trying to bisect or extract a subset of commits:

Screenshot 2024-07-30 at 10 23 21 AM


Forcing an entirely linear history (rebase *and* always squash) would avoid the tangled mess of that last case, but it would also lose frequently-useful semantic information: sometimes a PR really wants to be implemented in two or three or six commits, each of which makes a separate logical step of changes, and squashing all those together into a single commit would make life harder for future developers that want to understand the original programmer's thinking.

In some cases, force-rebase will untangle the cases I'm worried about, but we'd need to experiment to see which cases would remain. (I have a vague memory of using GitHub's "Rebase" button on a branch that had tangles but which was otherwise up-to-date, and I think it didn't feel the need to do anything, whereas using it when the branch was even one commit behind ended up cleaning things correctly).

Also, I know git has some tools to rewrite a branch and automatically act upon commits with `fixup!` or `amend!`, as a more-automated form of `git rebase --interactive`. I'd be totally happy with a force-rebase which executes those instructions (and I've seen plenty of tiny `fixup!` commits land on trunk when someone forgot to run that step before pushing. Is that what "autosquash" means in this context?

**Comment by frazarshad** [comment]
@warner the PR does not squash all the commits into a single commit. so all individual commits should be present in the commit history.
Just to make sure i ran it once again by merging this test [PR](https://github.com/frazarshad/mergify-experiements/pull/11) and here is the resulting commit history. It is the same as our ideal scenario
Screenshot 2024-07-30 at 9 52 49 PM

**Comment by frazarshad** [comment]
> I'm confused, does this force a rebase on every PR that is labelled `automerge:rebase`? That seems like that may be costly on PRs that don't have any fixup or merge commits.

@mhofman yes it forces a rebase on all PRs. this was done to ensure that the rebase action runs in every case scenario without fail. 
something we can do to avoid this is to add the condtion`-linear-history` to the rebase action. With this, it will only run when there is a merge commit in the history, in all other cases, we will be depending on our normal rebase (the one that occurs when the PR is enqueued in the merge queue) to fix the commit history.

> I also don't quite understand how the added condition for linear history helps triggering this 2, step merge process.

This condition was added to avoid any PRs from going into the merge queue before being rebased by the rebase action

**Comment by mhofman** [comment]
Oh I didn't realize `linear-history` was a new mergify condition, not the `linear-history` check that I wrote. While it's great to avoid merge commits, it doesn't address the concern of up to date PRs with `fixup!` commits, which our check enforces.

However, the `linear-history` check I wrote currently [does not run at all](https://github.com/Agoric/agoric-sdk/blob/4696a6ae369291fb492340d1444f3d205de49ece/.github/workflows/mergify-ready.yml#L51-L52) on `automerge:rebase` PRs.

Unless we ask mergify to implement an auto-squashable commit check, I think what we need is something like the following:
- extend out own `linear-history` check to run on `automerge:rebase`
- output a variable from that job if a rebase is needed
- have a dependent job that does nothing, but only executes if the above variable said a rebase was needed
- change the rebase merge queue condition to proceed if the above job is "skipped"
- trigger the new rebase action if that job has a "success" status

The main risk is an infinite rebase loop where `git rebase --autosquash` does not correctly fixup the commits (I have seen this happen, e.g. when the fixup commit references the wrong message). I am not sure how to guard against that.

**Comment by frazarshad** [comment]
>  it doesn't address the concern of up to date PRs with `fixup!` commits, which our check enforces.

@mhofman This is already handled by the when our `rebase updates then merge to master` rule rebases the PR before enqueueing it. I tested this out in this [PR](https://github.com/frazarshad/mergify-experiements/pull/14).

The `rebase updates then merge to master` rule handles rebasing when history is linear and `rebase and autosquash` handles the rebasing when the history is non-linear.

While this seperation of responsibilities between two rules is less than ideal, I think its more reliable than the alternative of writing our own CI job to handle this.

**Comment by mhofman** [comment]
> @mhofman This is already handled by the when our `rebase updates then merge to master` rule rebases the PR before enqueueing it.

To clarify, it does so because it forces a rebase of all `automerge:rebase` PRs, even if the PR already has an up to date linear history with no fixup commits?

Unless the rebase done by mergify does not change the commit in these cases (which would be wonderful), I think we can't afford a force rebase as it would force a re-run of all CI jobs when not necessary.

If the commit does not change, there may still be the case of erroneous `fixup!` commits that don't apply, or more generally failed rebase. I would like to know what the behavior of mergify is in these cases.

> I tested this out in this [PR](https://github.com/frazarshad/mergify-experiements/pull/14).

Thanks for that video. FYI, the `integration-test-result` check should not be GH required, so that the PR can embark the merge train early. I think you need to reconfigure your test repo. The merge-strategy-chosen however should be a required check.

**Comment by frazarshad** [comment]
> To clarify, it does so because it forces a rebase of all `automerge:rebase` PRs, even if the PR already has an up to date linear history with no fixup commits?

@mhofman no. fixup commits are already handled by the mergify config in master. this PR only adds the ability to handle merge commits on up to date PRs

> Unless the rebase done by mergify does not change the commit in these cases (which would be wonderful), I think we can't afford a force rebase as it would force a re-run of all CI jobs when not necessary.

As far as i have observed other than the case where there are no merge or fixup commits in a PR and the PR is up to date, the commit hashes are changed in all cases. This is true of our old way of doing things as well as our new method. In cases where the old method would not rebase a PR, the new method would not rebase a PR as well, hence the CI would not be re run

To make things more clear and understandable, i've tested every possible scenario i can come up with and compiled them in this [document](https://drive.google.com/file/d/1kfIW-pVOORuYwBOQlxt-2cus0k9ynAfb/view?usp=drive_link). Let me know if theres anything ive missed in this. 
Overall i think this two step rebasing is the best case scenario.

File: .github/workflows/check-fixup-commits.yml
Comment (mhofman) [inline .github/workflows/check-fixup-commits.yml:1]: Any reason not to add this as a job in `mergify-ready.yml` ?

Please also confirm that we cannot use the existing `linear-history` check in that workflow (after adding the `automerge:rebase` in the conditions).

From what I understand we still can't use it because as a required check it would prevent entering the queue at all, so for PRs with only fixup commits (no merge commit) the `linear-history` mergify condition would be true, but the queue conditions would fail (as a required check).

I suppose one approach would be to trigger the "rebase and autosquash" action when the `linear-history` check is failing, but I don't like that it would prevent the PR from entering the merge queue immediately (I understand that the rebase before entering the queue needs to happen if merge commits are present, and I'm ok with that because people in general shouldn't need to manually update their PR like that).

File: .github/workflows/check-fixup-commits.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      - ready_for_review
+      - labeled
+      - unlabeled
+      - auto_merge_enabled
+      - auto_merge_disabled
+
+jobs:
+  check-fixup-commits:
+    runs-on: ubuntu-latest
+
```
Comment (mhofman) [inline .github/workflows/check-fixup-commits.yml]: Let's only run this in case a rebase strategy has been chosen to avoid spamming people with failing tests notifications.

```suggestion
    if: >-
      github.event_name == 'pull_request' &&
      github.event.pull_request.draft == false &&
      github.event.pull_request.base.ref == 'master' && 
      contains(github.event.pull_request.labels.*.name, 'automerge:rebase') &&
      !contains(github.event.pull_request.labels.*.name, 'bypass:linear-history')
```

File: .github/workflows/check-fixup-commits.yml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      - synchronize
+      - converted_to_draft
+      - ready_for_review
+      - labeled
+      - unlabeled
+      - auto_merge_enabled
+      - auto_merge_disabled
+
+jobs:
+  check-fixup-commits:
```
Comment (mhofman) [inline .github/workflows/check-fixup-commits.yml]: Let's name this job/check after what it checks

```suggestion
  no-fixup-commits:
```

File: .mergify.yml [hunk 3]
```diff
@@ -21,6 +21,9 @@ queue_rules:
               - label=proto:expect-breakage
     merge_conditions:
       - base=master
+      - or:
+          - label=bypass:fixup
+          - check-success=check-fixup-commits
```
Comment (mhofman) [inline .mergify.yml]: Please add a comment that a merge condition is used to allow the PR to enter the queue in case the PR is already linear.

Also I would like to re-use the existing `bypass:linear-history` label instead. And if we add the conditions on the job as I suggested, we need to accept skipped checks as well

```suggestion
      # Rebase PRs with fixup commits are allowed to enter the merge queue but 
      # should not be allowed to merge if there are leftover fixup commits after rebase
      - or:
          - check-skipped=no-fixup-commits
          - check-success=no-fixup-commits
```

**Review by mhofman — COMMENTED** [review]
Some minor tweaks requested

File: .github/workflows/check-fixup-commits.yml
Comment (frazarshad) [reply .github/workflows/check-fixup-commits.yml:1]: > Please also confirm that we cannot use the existing linear-history check in that workflow (after adding the automerge:rebase in the conditions).

Im not exactly sure what you mean here. but if i understand correctly, you're talking about the `linear-history` job in `mergify-ready.yml` which i dont think this PR affects in any way. 

> but I don't like that it would prevent the PR from entering the merge queue immediately

If you think that the PR should enter the queue immediately, we can move the `linear-history` mergify check from the `pull_request_rules` to the `merge_conditions` in `queue_rules`. That way the PR can be rebased while it is in queue. but im not sure how the rebase of the `rebase and autosquash` action and `rebase updates then merge to master` will interact when they happen at the same time

File: .mergify.yml [hunk 4]
```diff
@@ -65,6 +77,7 @@ pull_request_rules:
     conditions:
       - base=master
       - label=automerge:rebase
+      - linear-history
```
Comment (mhofman) [inline .mergify.yml]: Thinking about this more, I think can get away with allowing non linear history PRs to enter the merge queue if they're not up to date

```suggestion
      - or:
        - '#commits-behind>0'
        - linear-history
```

File: .mergify.yml [hunk 5]
```diff
@@ -76,3 +89,11 @@ pull_request_rules:
     actions:
       queue:
         merge_method: squash
+  - name: rebase and autosquash
+    conditions:
+      - base=master
+      - label=automerge:rebase
+      - -linear-history
```
Comment (mhofman) [inline .mergify.yml]: similarly we should be able to restrict how often this runs to cases not covered by the above

```suggestion
      - '#commits-behind=0'
      - or:
        - -linear-history
        - check-failed=no-fixup-commits
      - -draft
```

File: .mergify.yml [hunk 6]
```diff
@@ -65,6 +77,7 @@ pull_request_rules:
     conditions:
       - base=master
       - label=automerge:rebase
+      - linear-history
     actions:
       queue:
         merge_method: merge
```
Comment (mhofman) [inline .mergify.yml:85]: I know this is the default, but should we specify `autosquash: true` here too to make things more obvious?

File: .github/workflows/mergify-ready.yml [hunk 7]
```diff
@@ -88,3 +88,32 @@ jobs:
...
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Check for fixup commits
+        id: fixup-commits
+        run: |
+          if [[ $(git rev-list "$BASE_SHA".."$HEAD_SHA" --grep="^fixup! " | wc -l) -eq 0 ]]; then
```
Comment (mhofman) [inline .github/workflows/mergify-ready.yml]: I just realized we're only checking `fixup!` commits here, and we should also check for `squash!` and `amend!` commits as well. As a drive-by, we should fix the `linear-history` job above to include `amend!` too

**Review by mhofman — COMMENTED** [review]
I realized we can avoid the separate rebase action in more cases.

File: .github/workflows/check-fixup-commits.yml
Comment (mhofman) [reply .github/workflows/check-fixup-commits.yml:1]: > we can move the `linear-history` mergify check from the `pull_request_rules` to the `merge_conditions` in `queue_rules`.

We can't since the `linear-history` I was referencing is the GH actions job, and is a "GH branch protection" required check, which is automatically transcribed into a `queue_condition` by mergify like every other GH required check (and should remain as such).

Because that check is required, it would prevent the non linear PR to enter the rebase queue, and so it wouldn't rebase. You're right that because of that we need a separate job that is not a GH required check, but is instead a queue merge condition.

**Comment by mhofman** [comment]
To summarize the requested change I think we'll end up with the following 3 cases when a non-draft PR has the `automerge:rebase` label:

- PR is up to date and linear, the PR will enter the merge queue
  - if the PR contains fixup commits, the `no-fixup-commit` check will fail, exit the queue, and trigger the rebase action
  - if the PR contains no fixup commits, it will simply merge
- PR is up to date and not linear, it will not enter the queue, but instead will trigger the rebase action.
- PR is not up to date (regardless of linear), it will enter the merge queue and immediately get rebased
  - after the rebase, see "PR is up to date and linear" (because a rebase makes a PR linear)

If I had time I'd make a diagram out of this.

File: .mergify.yml (hunk 6 above)
Comment (frazarshad) [reply .mergify.yml:85]: that field is deprecated and will probably wont be supported in a few months time like the other deprecated fields

**Review by mhofman — APPROVED** [review]
Thanks for addressing all these! Were you be able to redo a round of validation on mergify-experiments with the newest changes? I'm only 95% confident in my suggestions ;)

**Comment by frazarshad** [comment]
> Thanks for addressing all these! Were you be able to redo a round of validation on mergify-experiments with the newest changes? I'm only 95% confident in my suggestions ;)

i validated all the main cases. especially verifying the new suggested parts
